### ISSUE TITLE
[Bug] [ROCm] Router GEMM output should always be fp32, and the expert correction bias should not be cast to bf16

### ISSUE THREAD

**Issue by b8zhong** [sgl-project/sglang#34857]
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

Two related dtype issues on the ROCm/aiter MoE routing path. Both affect GLM-5.2. (1) is confirmed to cause precision issues; (2) is more minor but can still cause precision issues.

**1. The router GEMM returns bf16.** The router GEMM output should always be fp32 in DeepSeek architectures.

`python/sglang/srt/layers/rocm_linear_utils.py:14`

```python
def aiter_dsv3_router_gemm(hidden_states, weight):
    return tgemm.mm(hidden_states, weight.detach(), otype=hidden_states.dtype)  # -> bf16
```

Every other branch of `MoEGate.forward` returns fp32 — `dsv3_router_gemm(out_dtype=torch.float32)` (`deepseek_v2.py:538`) and `linear_bf16_fp32` (`deepseek_v2.py:548`). The aiter branch (`deepseek_v2.py:543`) is the only one that does not.

**2. The expert correction bias is cast to bf16, twice.**

At allocation, `python/sglang/srt/models/deepseek_v2.py:481`:

```python
correction_bias_dtype = torch.float32
if quant_config is not None:
    if _use_aiter and quant_config.get_name() in ("fp8", "compressed_tensors", "quark"):
        correction_bias_dtype = torch.bfloat16
```

The checkpoint stores it in fp32 (GLM-5.2: `mlp.gate.e_score_correction_bias  F32 [256]`), so this rounds it at load. Then again on every forward, `python/sglang/srt/layers/moe/topk.py:907` and `:1578`:

```python
correction_bias.to(dtype=gating_output.dtype)   # bf16 on aiter, per issue 1
```

The CUDA paths do the opposite — they upcast (`topk.py:925`, `topk.py:1473`) or hard-assert fp32 (`moe_fused_gate.py:60`).

#29783 fixed exactly this pair on the CUDA side: it dropped the `modelopt_fp4` + flashinfer-trtllm bf16 correction-bias cast, and switched the CUDA fallback from `F.linear` to `linear_bf16_fp32` so the output is fp32. The aiter branch was left as-is in that PR.

**Comment by b8zhong** [comment]
Hi @billishyahao, could you help take a look at this issue, as currently it will cause accuracy differences between ROCm and CUDA. Thanks.
